Add lint about redefining runtime symbols

Urgau opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/146505)*
<!-- homu-ignore:end -->

This PR adds lint to warn about redefinition of runtime symbols[^name] that are assumed and used by `core`[^1] and `std`.

We have had multiple reports of users tripping over this:
 - [Why does `#[no_mangle] fn open() {}` make `cargo t` hang?](https://users.rust-lang.org/t/why-does-no-mangle-fn-open-make-cargo-t-hang/103423)
 - [Pointer becomes misaligned in test with `no_mangle`](https://users.rust-lang.org/t/pointer-becomes-misaligned-in-test-with-no-mangle/126580)

## `redefining_runtime_symbols`

*Old proposed name: `clashing_function_names_with_fundamental_functions`*

*(warn-by-default)*

The `redefining_runtime_symbols` lint checks for items whose symbol name redefines a runtime symbols expected by `core` and/or `std`.

### Example

```rust,no_run
#[unsafe(no_mangle)]
pub fn strlen() {} // redefines the libc `strlen` function
```

```text
warning: redefinition of the runtime `strlen` symbol used by the standard library
 --> a.rs:2:1
  |
2 | pub fn strlen() {}
  | ^^^^^^^^^^^^^^^^^^
  |
  = note: extra care must be taken when redefining those symbols, they must match exactly (ABI, function arguments, function return type, behavior, ...)
  = note: see <https://doc.rust-lang.org/core/index.html#how-to-use-the-core-library> for the more details
  = help: either allow this lint or remove any `#[unsafe(no_mangle)]` or `#[unsafe(export_name = "strlen")]`
  = note: `#[warn(redefining_runtime_symbols)]` on by default
```

### Explanation

Up-most care is required when redefining runtime symbols assumed and used by the standard library. They must follow the C specification, not use any standard-library facility or undefined behavior may occur.

The symbols currently checked are respectively:
 - from `core`[^1]: `memcpy`, `memmove`, `memset`, `memcmp`, `bcmp`, `strlen`
 - from `std`: `open`/`open64`, `read`, `write`, `close`

[^1]: https://doc.rust-lang.org/core/index.html#how-to-use-the-core-library

@rustbot labels +I-lang-nominated +T-lang +needs-fcp +A-lints
cc @traviscross
r? compiler

[^name]: previous lint name `clashing_function_names_with_fundamental_functions`, bike-shed at https://github.com/rust-lang/rust/pull/146505#issuecomment-3288716835

Urgau committed:
Add lint warn about clashing function names with fundamental functions

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17696313973/job/50296291620) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50296291620) [(plain)](https://github.com/rust-lang/rust/commit/99437123354bc79990a489ce49afe49972df108f/checks/50296291620/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- compiler/rustc_lint/src/fundamental_functions.rs - fundamental_functions::CLASHING_FUNCTION_NAMES_WITH_FUNDAMENTAL_FUNCTIONS (line 13) stdout ----
error: cannot find attribute `undafe` in this scope
 --> compiler/rustc_lint/src/fundamental_functions.rs:14:3
  |
3 | #[undafe(no_mangle)]
  |   ^^^^^^

error: aborting due to 1 previous error

Couldn't compile the test.
```

</details>

rust-timer commented:
Queued fb73ebde4cd6ae9fb27bac8017eab9dc519131f9 with parent 064cc81354a940e297a1be4dfa9e26759c8431be, future [comparison URL](https://perf.rust-lang.org/compare.html?start=064cc81354a940e297a1be4dfa9e26759c8431be&end=fb73ebde4cd6ae9fb27bac8017eab9dc519131f9).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (fb73ebde4cd6ae9fb27bac8017eab9dc519131f9): [comparison URL](https://perf.rust-lang.org/compare.html?start=064cc81354a940e297a1be4dfa9e26759c8431be&end=fb73ebde4cd6ae9fb27bac8017eab9dc519131f9&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=064cc81354a940e297a1be4dfa9e26759c8431be&end=fb73ebde4cd6ae9fb27bac8017eab9dc519131f9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=064cc81354a940e297a1be4dfa9e26759c8431be&end=fb73ebde4cd6ae9fb27bac8017eab9dc519131f9&stat=max-rss)
<details>
<summary>Results (secondary -0.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 8.2%  | [8.2%, 8.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.7% | [-2.5%, -1.1%] | 6     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=064cc81354a940e297a1be4dfa9e26759c8431be&end=fb73ebde4cd6ae9fb27bac8017eab9dc519131f9&stat=cycles:u)
<details>
<summary>Results (secondary -2.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-3.6%, -1.6%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=064cc81354a940e297a1be4dfa9e26759c8431be&end=fb73ebde4cd6ae9fb27bac8017eab9dc519131f9&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 468.794s -> 471.13s (0.50%)
**Artifact size**: 388.08 MiB -> 388.08 MiB (0.00%)

Urgau committed:
Rename lint to `redefining_runtime_symbols`

Urgau commented:
> Perhaps `redefining_runtime_symbols` (or maybe `redefined_runtime_symbols`) would be a decent name?

`redefining_runtime_symbols` works for me. Changed the lint name as such.

> Also, shouldn't we be checking for more than just functions?

Indeed, forgot that statics also have a symbol. Fixed.

---------

Regarding the lint level, I made it warn-by-default since there are legitimate reasons to implement those symbols (like when implementing a libc), but maybe it should be deny-by-default?

bjorn3 commented on compiler/rustc_lint/src/redefining_runtime_symbols.rs:
```diff
@@ -0,0 +1,72 @@
+
+        // Compute the symbol name of our item (without mangling, as our mangling cannot ever
+        // conflict with runtime symbols).
+        let Some(symbol_name) = rustc_symbol_mangling::symbol_name_without_mangling(
+            cx.tcx,
+            rustc_middle::ty::InstanceKind::Item(item.owner_id.to_def_id()),
+        ) else {
+            return;
+        };
```
This could just be

```rust
        if let Some(name) = attrs.symbol_name {
            // Use provided name
            return name.to_string();
        }

        if attrs.flags.contains(CodegenFnAttrFlags::NO_MANGLE) {
            // Don't mangle
            return tcx.item_name(def_id).to_string();
        }
```

right? The rest of `symbol_name_without_mangling` is not relevant here.

Urgau replied:
Technically yes, but I don't want the logic to be at two different places if we don't have to, and since the perf run is clean as is, I would like to keep `symbol_name_without_mangling` (to avoid duplicating the logic).

bjorn3 replied:
The function name is not entirely correct given that aside from this snippet, all code inside this function handles cases where there actually is symbol mangling because of `#[rustc_std_internal_symbol]`, because of `#[wasm_import_module]` or because it is the static containing all defined proc macros.

Urgau committed:
Add `open64` and rename `symbol_name_without_mangling`

scottmcm commented:
Wishlist: it'd be really nice if rustc looked at the things that you are `extern fn`ing or `no_mangle`ing and checking to make sure that those are consistent with anything that your dependencies are doing.

joshtriplett commented:
One further note: once we have the `deny` case for having the wrong type, we should evaluate whether the `warn` case is catching more accidents or catching mostly people who are doing this *intentionally*. If the latter, we may wish to change it from `warn` to `allow` or drop it.

bors commented:
:umbrella: The latest upstream changes (presumably #149222) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Urgau commented:
I've opened https://github.com/rust-lang/rust/pull/155521 which replaces this attempt.
